torch/distributed/fsdp/optim_utils.py

```diff
@@ -488,16 +607,261 @@ def _flatten_non_tensor_optim_state(
+
+    Returns:
+        Dict[str, Any]: The optimizer state dict with the positive-dimension
+        tensor state correctly populated via ``broadcast()`` s from rank 0.
+    """
+    assert rank != 0 or flat_optim_state_dict is not None, \
+        "Expects rank 0 to pass in the flattened optimizer state dict"
+    no_tensor_osd = processed_optim_state_dict  # alias
+    flat_osd = flat_optim_state_dict  # alias
+    for param_id, param_state in no_tensor_osd["state"].items():
+        for state_name, value in param_state.items():
+            is_pos_dim_tensor_state = isinstance(value, _PosDimTensorInfo)
+            if not is_pos_dim_tensor_state:
+                continue
+            if rank == 0:
+                assert flat_osd is not None
+                unsharded_tensor = flat_osd["state"][param_id][state_name]
+            else:
+                unsharded_tensor = None
+            shape, dtype = value.shape, value.dtype
+            if param_id in fsdp_flat_param_ids:  # FSDP parameter
+                _broadcast_sharded_pos_dim_tensor_state(
+                    unsharded_tensor, param_state, state_name, shape, dtype,
+                    broadcast_device, rank, world_size, group,
+                )  # modify `param_state` destructively
+            else:  # non-FSDP parameter
+                _broadcast_unsharded_pos_dim_tensor_state(
+                    unsharded_tensor, param_state, state_name, shape, dtype,
+                    broadcast_device, rank, group,
+                )  # modify `param_state` destructively
+    return no_tensor_osd
+
+
+def _broadcast_sharded_pos_dim_tensor_state(
+    unsharded_tensor: Optional[torch.Tensor],
+    param_state: Dict[str, Any],
+    state_name: str,
+    shape: torch.Size,
+    dtype: torch.dtype,
+    broadcast_device: torch.device,
+    rank: int,
+    world_size: int,
+    group,
+) -> None:
+    """
+    Broadcasts positive-dimension tensor state for the state ``state_name``
+    corresponding to an FSDP parameter shard-by-shard, only to be saved on the
+    relevant rank. This modifies ``param_state`` destructively.
+
+    Args:
+        unsharded_tensor (Optional[torch.Tensor]): Unsharded tensor from which
+            to broadcast shards if on rank 0; ignored otherwise.
+        shape (torch.Size): Shape of the sharded tensor; same on all ranks.
+    """
+    get_shard: Optional[functools.partial[Tuple[torch.Tensor, int]]] = None
+    if rank == 0:
+        assert unsharded_tensor is not None, \
+            "Expects rank 0 to pass in the unsharded tensor"
+        unsharded_tensor = unsharded_tensor.to(broadcast_device)
+        get_shard = functools.partial(
+            FSDP.FullyShardedDataParallel._get_shard_functional,
+            unsharded_tensor,
+        )
+    for target_rank in range(1, world_size):
+        if rank == 0:
+            assert get_shard is not None
+            sharded_tensor = get_shard(target_rank, world_size)[0].to(broadcast_device)
```

Comment:
nit: looks like no need to call .to(broadcast_device) as unsharded_tensor has already been moved? maybe I'm wrong

Reply:
Good catch. I fixed this so that only the shard is moved to `broadcast_device`. Otherwise, the peak GPU memory usage on rank 0 is possibly increased.